activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -171,6 +172,14 @@ def initialize(config_or_deprecated_connection, deprecated_logger = nil, depreca
         @verified = false
       end
 
+      def synchronized=(synchronized) # :nodoc:
+        @lock = if synchronized
+          ActiveSupport::Concurrency::LoadInterlockAwareMonitor.new
+        else
+          ActiveSupport::Concurrency::NullLock
```

Comment:
Would it make sense to just have `@lock` be nil for the non-synchronized case? You’d have to change all usages to `@lock&.synchronize` but I think it’s more clear that there is no lock sometimes and prevents us from having a null lock. 🤷

Reply:
🤦 You're right. I meant something like this to get rid of the NullLock:

```ruby
def synchronize(&block)
  synchronized ? @lock.synchronize(&block) : yield
end
```

Then call `synchronize` everywhere instead of `@lock.synchronize`

Reply:
Sounds good!